app/controllers/admin/api/objects_controller.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+class Admin::Api::ObjectsController < Admin::Api::BaseController
+
+  class AllowedStatusObject
+    attr_reader :controller, :status_object
+
+    def initialize(controller, status_object)
+      @controller = controller
+      @status_object = status_object
+    end
+
+    def authorize
+      raise NotImplementedError
+    end
+
+    private
+
+    def raise_access_denied
+      raise CanCan::AccessDenied
+    end
+  end
+
+  class BuyerAccountObject < AllowedStatusObject
+    def authorize
+      controller.authorize! :read, status_object
+    end
+  end
+
+  class ServiceObject < AllowedStatusObject
+    def authorize
+      raise_access_denied unless controller.accessible_services.exists?(service_id)
+    end
+
+    def service_id
+      controller.status_object_id
+    end
+  end
+
+  class ProxyObject < ServiceObject
+    delegate :service_id, to: :status_object
+  end
+
+  class BackendApiObject < AllowedStatusObject
+    def authorize
+      controller.provider_can_use!(:api_as_product)
```

Comment:
We have a cancan ability for this which is what we are using in the BackendApisController (both in UI and API), so maybe we should do:
```suggestion
      controller.authorize! :read, status_object
```